### PR TITLE
rewrite state handling, phase 1+2

### PR DESCRIPTION
**Pull request by warner**
This rewrites the kernel state management, in support of #144. Previously, kernel/vat state was stored in a composite object, with everything being JSON-serialized into a single big string at the end of the batch/block. In addition, the kernel promise-management code was a bit casual about mutating that state from outside the keepers.

In phase 1, we still use JSON-serialization, but the object is now reduced to a strict string/string key-value store, and the storage/keeper layer returns hardened objects (so all mutations must go through the storage API).

In phase 2 (also in this PR), the internal API is refactored to reflect the upcoming DB interface. The JSON-serialized KV object is wrapped in a has/get/set -style API.

This PR retains compatibility with the existing host API (i.e. `buildVatController()` still accepts an `initialState=` string, and still has a `.getState()` method that returns a string). The next phases will replace this with a host-provided DB object.

### CODE REVIEW

File: src/kernel/state/kernelKeeper.js [hunk 1]
```diff
@@ -225,61 +332,78 @@ function makeKernelKeeper(initialState) {
   }
 
   function getAllVatIDs() {
-    return Array.from(Object.getOwnPropertyNames(state.vats)).sort();
+    const nextID = Nat(Number(storage.get(`vat.nextID`)));
+    const vatIDs = [];
+    for (let i = 1; i < nextID; i += 1) {
```
Comment (gamedevsam) [inline src/kernel/state/kernelKeeper.js]: Why does this iterator start at 1?

File: src/kernel/state/kernelKeeper.js [hunk 2]
```diff
@@ -335,26 +459,23 @@ function makeKernelKeeper(initialState) {
...
 
-    const { kernelPromises } = state;
-    Object.getOwnPropertyNames(kernelPromises).forEach(s => {
-      const kp = { id: s };
-      const p = kernelPromises[s];
-      Object.defineProperties(kp, Object.getOwnPropertyDescriptors(p));
-      if (p.subscribers) {
-        kp.subscribers = Array.from(p.subscribers);
+    const nextPromiseID = Nat(Number(storage.get('kp.nextID')));
+    for (let i = 0; i < nextPromiseID; i += 1) {
```
Comment (gamedevsam) [inline src/kernel/state/kernelKeeper.js]: But it starts at 0 here?

File: src/kernel/state/kernelKeeper.js [hunk 3]
```diff
@@ -225,61 +332,78 @@ function makeKernelKeeper(initialState) {
...
       ephemeral.deviceKeepers.set(deviceID, dk);
     }
     return ephemeral.deviceKeepers.get(deviceID);
   }
 
   function getAllDeviceIDs() {
-    return Array.from(Object.getOwnPropertyNames(state.devices)).sort();
+    const nextID = Nat(Number(storage.get(`device.nextID`)));
+    const deviceIDs = [];
+    for (let i = 1; i < nextID; i += 1) {
```
Comment (gamedevsam) [inline src/kernel/state/kernelKeeper.js]: Same here, why `i = 1` instead of `i = 0`?

**Comment by warner** [comment]
Good questions. I'll replace those starting points with better-defined constants in the next phase of changes. The first one (that counts up to `vat.nextID`) should start with the same value that `vat.nextID` starts with in `createStartingKernelState()`: 1. The second one (`kp.nextID`) should have started at 40, etc. Thanks for the catch!

**Comment by bhgsbatista** [comment]
Totally out of my element here, but very happy my curiosity caught a bug, glad I asked :)

**Pull request merged**